evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,161 @@
 import logging
 
+from cytoolz import (
+    pipe,
+)
+
+import rlp
+
 from web3.contract import (
     Contract,
 )
 
 from eth_utils import (
+    event_signature_to_log_topic,
     is_canonical_address,
     to_checksum_address,
     to_dict,
+    to_tuple,
+)
+
+from evm.rlp.sedes import (
+    address,
+    hash32,
 )
 
 from evm.utils.hexadecimal import (
+    encode_hex,
     decode_hex,
 )
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
 
 from evm.vm.forks.sharding.config import (
     get_sharding_config,
 )
 
 
+class NextLogUnavailable(Exception):
+    pass
+
+
+@to_dict
+def parse_collation_added_data(data_hex):
+    data_bytes = decode_hex(data_hex)
+    score = big_endian_to_int(data_bytes[-32:])
+    is_new_head = bool(big_endian_to_int(data_bytes[-64:-32]))
+    header_bytes = data_bytes[:-64]
+    # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+    sedes = rlp.sedes.List([
+        rlp.sedes.big_endian_int,
+        rlp.sedes.big_endian_int,
+        hash32,
+        hash32,
+        hash32,
+        address,
+        hash32,
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
     def __init__(self, *args, default_privkey, **kwargs):
         self.default_privkey = default_privkey
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         self.config = get_sharding_config()
+        # TODO: currently set one log_handler for each shard. Should see if there is a better way
+        #       to make one log_handler shared over all shards.
+        self.log_handlers = {}
+        # shard_id -> list
+        # older <---------------> newer
+        self.new_logs = {}
+        # shard_id -> list
+        self.unchecked_logs = {}
+        # shard_id -> score
+        self.current_score = {}
+
         super().__init__(*args, **kwargs)
 
+    def init_log_variables(self, shard_id):
+        self.new_logs[shard_id] = []
+        # shard_id -> list
+        self.unchecked_logs[shard_id] = []
+        # shard_id -> score
+        self.current_score[shard_id] = None
```

Comment:
I think that this entire class would be cleaned up by implementing one more data structure to house all of the shard data.

```python
class ShardTracker:
    shard_id = ...
    log_handler = ...
    new_logs = ...
    unchecked_logs = ...
    current_score = ...
```

I suspect that some, if not most of the business logic for updating the shard state could then be moved into that object at which point this class would just manage all of the individual instances of that class.

Thoughts?  I think it would clean this class up in a significant way.

Reply:
I think that is a good idea. I will do it.
Just two questions about it.
1. Should I put `ShardTracker` in another file `shard_tracker.py`?
2. Since all logics regarding shards logs are moved to `ShardTracker`(including `get_next_log`, `fetch_candidate_head`), if there is a need for an API in `VMCHandler` like `fetch_candidate_head`, may I just set a same name method `fetch_candidate_head` in `VMCHandler` and simply calls `ShardTracker.fetch_candidate_head`? (Though I think there might only be new methods like `guess_head` in VMC which will be implemented in the future)
Very much appreciate!

Reply:
Moreover, may I request a bumpversion in `web3.py`, to resolve the error in CI?
Thank you!

Reply:
1. It's probably fine to start with it here in this file or to put it in it's own module.  Up to you.  `shard_tracker.py` seems like a fine choice as well.

2. If this makes `VMCHandler` obsolete and you think you're able to do what you need with only the `ShardTracker` class, then it's probably a good idea to do away with it.  Really depends on how these APIs will be used and whether you need an abstraction to track *many* shards and whether it still makes sense for that to be a class.

Reply:
@pipermerriam Thank you for the great suggestions! I have moved those logics to `ShardTracker`, currently in the same module `vmc_handler` as `VMCHandler`.
For now I just make `get_next_log`, `fetch_candidate_head` directly calls the corresponding `shard_tracker`. I think I can determine what to remove and how to restructure between `ShardTracker` and `VMCHandler` after `GUESS_HEAD` and `CREATE_COLLATION` are implemented, which will be done in the next stage.